scoutcloud/scoutcloud/src/logic/jobs/balance.rs

```diff
@@ -0,0 +1,272 @@
+use scoutcloud_entity as db;
+use sea_orm::{
+    prelude::*, ActiveValue::Set, DbBackend, FromQueryResult, IsolationLevel, Statement,
+    TransactionTrait,
+};
+use std::ops::Mul;
+use tracing::instrument;
+
+#[derive(fang::serde::Serialize, fang::serde::Deserialize)]
+#[serde(crate = "fang::serde")]
+pub struct CheckBalanceTask {
+    schedule: Option<String>,
+    #[cfg(test)]
+    database_url: Option<String>,
+}
+
+impl_get_db!(CheckBalanceTask);
+
+impl Default for CheckBalanceTask {
+    fn default() -> Self {
+        Self {
+            schedule: Some("0 * * * * *".to_string()),
+            #[cfg(test)]
+            database_url: None,
+        }
+    }
+}
+
+#[typetag::serde]
+#[fang::async_trait]
+impl AsyncRunnable for CheckBalanceTask {
+    #[instrument(err(Debug), skip(self, client), level = "info")]
+    async fn run(&self, client: &mut dyn AsyncQueueable) -> Result<(), FangError> {
+        let db = self.get_db().await;
+        // 'check balance' is unique task and there is only one instance of this task running
+        // however we begin transaction with serializable isolation level just in case
+        let tx = db
+            .begin_with_config(Some(IsolationLevel::Serializable), None)
+            .await
+            .map_err(DeployError::Db)?;
+        let unpaid_list = UnpaidDeployment::all(&tx).await.map_err(DeployError::Db)?;
+        if unpaid_list.is_empty() {
+            return Ok(());
+        }
+        tracing::info!(unpaid = ?unpaid_list, "found {} unpaid deployments. start processing", unpaid_list.len());
+        for unpaid in unpaid_list {
+            if unpaid.creator_can_pay(&tx).await.map_err(DeployError::Db)? {
+                unpaid.mark_as_paid(&tx).await.map_err(DeployError::Db)?;
+            } else {
+                tracing::warn!(
+                    user_id = unpaid.creator_id,
+                    deployment_id = unpaid.deployment_id,
+                    "user can't pay for deployment. stopping deployment",
+                );
+                // TODO: maybe notify user?
+                client
+                    .insert_task(&StoppingTask::from_deployment_id(unpaid.deployment_id))
+                    .await?;
+            }
+        }
+        tx.commit().await.map_err(DeployError::Db)?;
+        Ok(())
+    }
+
+    fn uniq(&self) -> bool {
+        true
+    }
+
+    fn cron(&self) -> Option<Scheduled> {
+        self.schedule
+            .as_ref()
+            .map(|s| Scheduled::CronPattern(s.clone()))
+    }
+}
+
+#[derive(Debug, FromQueryResult, PartialEq, Eq)]
+struct UnpaidDeployment {
+    deployment_id: i32,
+    total_used_hours: i32,
+    total_paid_hours: i32,
+    cost_per_hour: Decimal,
+    creator_id: i32,
+}
+
+impl UnpaidDeployment {
+    pub fn hours(&self) -> i32 {
+        self.total_used_hours - self.total_paid_hours
+    }
+
+    pub fn expense_amount(&self) -> Decimal {
+        self.cost_per_hour.mul(Decimal::new(self.hours() as i64, 0))
```

Comment:
nit: use `saturating_mul`; it does not panic
(it seems that it's not possible in the case, thus nit c:)

Reply:
well, if we have situation when `cost_per_hour * price` should panic, I want my code to panic. not saturating the result and pretending that everything is ok